lib/models/message.cpp

```diff
@@ -0,0 +1,151 @@
+#include <models/message.hpp>
+
+namespace disccord
+{
+    namespace models
+    {
+        message::message()
+            : author(), id(""), channel_id(""), content(""),
+            timestamp(""), edited_timestamp(), nonce(),
+            webhook_id(), tts(false), mention_everyone(false),
+            pinned(false), mentions(), mention_roles(),
+            attachments(), embeds(), reactions()
+        { }
+
+        message::~message()
+        { }
+
+        void message::decode(web::json::value json)
+        {
+            id = json.at("id").as_string();
+            channel_id = json.at("channel_id").as_string();
+            content = json.at("content").as_string();
+            timestamp = json.at("timestamp").as_string();
+            tts = json.at("tts").as_bool();
+            mention_everyone = json.at("mention_everyone").as_bool();
+            pinned = json.at("pinned").as_bool();
+            
+            #define get_field(var, conv) \
+                if (json.has_field(#var)) { \
+                    auto field = json.at(#var); \
+                    if (!field.is_null()) { \
+                        var = decltype(var)(field.conv()); \
+                    } else { \
+                        var = decltype(var)::no_value(); \
+                    } \
+                } else { \
+                    var = decltype(var)(); \
+                }
+                
+            #define get_composite_field(var, type) \
+                if (json.has_field(#var)) { \
+                    auto field = json.at(#var); \
+                    if (!field.is_null()) { \
+                        type val; \
+                        val.decode(field); \
+                        var = decltype(var)(val); \
+                    } else { \
+                        var = decltype(var)::no_value(); \
+                    } \
+                } else { \
+                    var = decltype(var)(); \
+                }
+
+            #define get_composite_field_vector(var, type) \

```

Comment:
a lot of composite `std::vector`s in this model, thus, a new macro was born! Also because of this, I have another one: `encode_composite_vector` macro down in the encode method